libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountBuilder.kt

```diff
@@ -0,0 +1,529 @@
+            .cookieClientSrc(tr.cookieClientSrc)
+            .cookieSidClient(tr.cookieSidClient)
+            .sidCookieName(tr.sidCookieName)
+    }
+
+    /**
+     * Set fields from id service response
+     * @param id id service response
+     * @return Instance of this class
+     */
+    fun populateFromIdServiceResponse(id: IdServiceResponse?): UserAccountBuilder {
+        if (id == null) {
+            return this
+        }
+
+        return username(id.username)
+            .firstName(id.firstName)
+            .lastName(id.lastName)
+            .displayName(id.displayName)
+            .email(id.email)
+            .photoUrl(id.pictureUrl)
+            .thumbnailUrl(id.thumbnailUrl)
+            .language(id.language)
+            .locale(id.locale)
+    }
+
+    /**
+     * Set fields from an existing UserAccount
+     * @param userAccount user account.
+     * @return Instance of this class
+     */
+    fun populateFromUserAccount(userAccount: UserAccount): UserAccountBuilder {
+        return this
+            .authToken(userAccount.authToken)
+            .refreshToken(userAccount.refreshToken)
+            .loginServer(userAccount.loginServer)
+            .idUrl(userAccount.idUrl)
+            .instanceServer(userAccount.instanceServer)
+            .orgId(userAccount.orgId)
+            .userId(userAccount.userId)
+            .username(userAccount.username)
+            .accountName(userAccount.accountName)
+            .communityId(userAccount.communityId)
+            .communityUrl(userAccount.communityUrl)
+            .firstName(userAccount.firstName)
+            .lastName(userAccount.lastName)
+            .displayName(userAccount.displayName)
+            .email(userAccount.email)
+            .photoUrl(userAccount.photoUrl)
+            .thumbnailUrl(userAccount.thumbnailUrl)
+            .additionalOauthValues(userAccount.additionalOauthValues)
+            .lightningDomain(userAccount.lightningDomain)
+            .lightningSid(userAccount.lightningSid)
+            .vfDomain(userAccount.vfDomain)
+            .vfSid(userAccount.vfSid)
+            .contentDomain(userAccount.contentDomain)
+            .contentSid(userAccount.contentSid)
+            .csrfToken(userAccount.csrfToken)
+            .nativeLogin(userAccount.nativeLogin)
+            .language(userAccount.language)
+            .locale(userAccount.locale)
+            .cookieClientSrc(userAccount.cookieClientSrc)
+            .cookieSidClient(userAccount.cookieSidClient)
+            .sidCookieName(userAccount.sidCookieName)
+            .clientId(userAccount.clientId)
+    }
+
+    /**
+     * Sets the allow unset behavior
+     * When false:
+     * - passing null to a setter does not unset a value previously set
+     * - passing a map to additionAuthValues causes the new map and old map to be merged
+     * @param allowUnset new value for allowUnset
+     */
+    fun allowUnset(allowUnset: Boolean): UserAccountBuilder {
```

Comment:
I believe the refresh token response should have the same fields as the login token response. But if it has a few missing fields, it should not cause them to be gone. The old code has some checks in some places, it would not blindly set some fields with the fields coming from the token response, it would check that there is actually a value.